README.md
<p align="center">
<img src="./assets/banner.png" /><br></p>

# CryptoBeacon - Annonymous and transparent decentralized platform for inclusive communities

_This PoC is developed as a solution for the MLH Neighborhood Hackathon (14-16 May 2021_)

## PoC Progress for MVP

- [x] Develop a smart contract with basic functionalities
- [x] Setup Dev environment and test locally
- [x] Deploy on Polygon Mumbai Testnet ( Deployed Contract [Here](https://explorer-mumbai.maticvigil.com/address/0xfe87576951D8c0d0c33Fd3053Cc586a461BD8c5E/contracts))
- [x] Working Web3
- [x] Set up initial front-end structure
- [ ] Build working Web3 for Front-end structure
- [ ] Connect Web3 with Deployed contract
- [ ] Set up and integrate IPFS
- [ ] Deploy Dapp Frontend on netlify

## Overview

**CryptoBeacon** is a community platform where people from all over the world can come to share their speeches and report the things happening concurrently. The data is secured and communicated through a **blockchain network**, a peer-to-peer decentralized technology coupled with a crypto storage facility. This allows people to overcome local restrictions imposed on internet connectivity and omits the need of using a _Virtual Private Network_ to access state-blocked contents and resources.

CryptoBeacon's two main objectives are:

- Estatablish a seamless platform to amplify community rights
- Provide annonymity and overcome state-level censorships

## Architecture

CryptoBeacon leverages the advantages of [Polygon Network](https://matic.network), which is a blockchain decentralized application platform that provides hybrid _proof-of-stake_ and _plasma-enabled_ sidechains.

The system creates a bridge between distributed file storage system powered by **IPFS (Inter Planetary File System)**.

## Features

- Annonymously publish a activity update on a decentralised app via wallet
- Ciew latest updates/news from the platform without any restrictions
- Each activity update can be Upvoted by the community to establish trust

## How it works

<img src="assets\system-arch.png" /><br>

The decentralized aspect is designed using **Ethereum** based scalable and efficient solution - **Polygon**.
We have used **IPFS**, which is a secured peer-to-peer network file storage. It helped us to remove single points of failure in our developed system. Combining all these, we have built this inclusive solution.

## Key benefits

- Overcome censorship and domain ban
- Resistant to most traditional web application attacks
- No VPN required
- No single point of failure

## :hammer: Technology Stack

1. **Blockchain Tools**

   1. [Truffle](https://www.trufflesuite.com/) (Framework)
   1. [Web3 JS](https://web3js.readthedocs.io/en/v1.2.11/) (Interface)

1. **Smart Contract Programming**
   1. [Solidity](https://solidity.readthedocs.io/en/v0.7.0/)
1. **Front End**

   1. [HTML](https://en.wikipedia.org/wiki/HTML/)
   1. [CSS](https://www.w3.org/Style/CSS/Overview.en.html)
   1. [Bootstrap](https://getbootstrap.com/)
   1. [JavaScript](https://www.javascript.com/)

1. **Dev Servers**

   1. [Lite-Server](https://www.npmjs.com/package/lite-server)
   1. [Ganache](https://www.trufflesuite.com/ganache)

1. **Distributed Networks**
   1. [Polygon](https://polygon.technology/)
   1. [IPFS](https://ipfs.io/)

## Our team

- [Anjum Rashid](https://github.com/bijoy26) (Project Manager) <br>
- [Jonayed Tanjim](https://github.com/tanjim01) (Resource Manager) <br>
- [Md. Asifuzzaman](https://github.com/asif-uz-zaman) (Content Creator) <br>
- [Kaiser Ahmed](https://github.com/mekaiser) (UI Designer) <br>


changelog.md
- [x] Install truffle globally `npm install -g truffle`
- [x] `truffle init`
- [x] `truffle develop`
- [x] `compile`
- [x] `migrate`
- [x] `npm run dev`
      Setups

## Error Logs (Truffle instllatoion)

msvs_version not set from command line or npm config
VCINSTALLDIR not set, not running in VS Command Prompt checking VS2019 (16.8.30907.101) found at:
"C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise"

- "Visual Studio C++ core features" missing
- could not find a version of Visual Studio 2017 or newer to use
- looking for Visual Studio 2015
- not found

not looking for VS2013 as it is only supported up to Node.js 8

---

You need to install the latest version of Visual Studio
including the "Desktop development with C++" workload.
For more information consult the documentation at:
https://github.com/nodejs/node-gyp#on-windows

---

stack Error: Failed to execute 'C:\Program Files\nodejs\node.exe C:\Users\Anjum\AppData\Roaming\nvm\v14.15.0\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js configure --fallback-to-build --module=C:\Program Files\nodejs\node_modules\truffle\node_modules\sqlite3\lib\binding\node-v83-win32-x64\node_sqlite3.node --module_name=node_sqlite3 --module_path=C:\Program Files\nodejs\node_modules\truffle\node_modules\sqlite3\lib\binding\node-v83-win32-x64 --napi_version=7 --node_abi_napi=napi --napi_build_version=0 --node_napi_label=node-v83' (1)
stack at ChildProcess.<anonymous> (C:\Users\Anjum\AppData\Roaming\nvm\v14.15.0\node_modules\truffle\node_modules\\lib\util\compile.js:83:29)
stack at ChildProcess.emit (events.js:315:20)
stack at maybeClose (internal/child_process.js:1048:16)
stack at Process.ChildProcess.\_handle.onexit (internal/child_process.js:288:5)

///////// GET DEPLOYED INSTANCE /////////

let cb = await Cryptobeacon.deployed();
let acc = await web3.eth.getAccounts();

let rec = await Cryptobeacon.getBalance(accounts[1])

///////// ADD FIRST ACTIVITY /////////

let t1 = "hi";
let d1 = "hi there people";
let c1 = "social";

cb.addActivity(web3.utils.utf8ToHex(t1),web3.utils.utf8ToHex(d1), web3.utils.utf8ToHex(c1))

///////// ADD SECOND ACTIVITY /////////

let t2 = "bob";
let d2 = "bob the builder";
let c2 = "admin";

cb.addActivity(web3.utils.utf8ToHex(t2),web3.utils.utf8ToHex(d2), web3.utils.utf8ToHex(c2))

///////// ADD THIRD ACTIVITY /////////

cb.addActivity(t2,d2,c2)

///////// GET ACTIVITY /////////

a = await cb.getActivity(1)

let act = await cb.getActivity(1)

Cryptobeacon.deployed().then(function(instance){return instance.getActivity(1);}).then(function(value){let a = value; console.log(`CID: ${a.cid}\n$Title: ${a.title}\n$details: ${a.details}\n$category: ${a.category}\n$Upvotes: ${a.upvotes}`)})

///////// UPVOTE /////////

cb.upVote(1)
cb.upVote(4)
///////// GET 5 ACTIVITIES /////////

a = await cb.getActivity(1)

let act = await cb.getActivity(1)

Cryptobeacon.deployed().then(function(instance){return instance.getLast5Activities();}).then(function(value){let a = value; console.log(`\nCID: ${a[0].cid}\n$Title: ${a[0].title}\n$details: ${a[0].details}\n$category: ${a[0].category}\n$Upvotes: ${a[0].upvotes}`);console.log(`\nCID: ${a[1].cid}\n$Title: ${a[1].title}\n$details: ${a[1].details}\n$category: ${a[1].category}\n$Upvotes: ${a[1].upvotes}`);console.log(`\nCID: ${a[2].cid}\n$Title: ${a[2].title}\n$details: ${a[2].details}\n$category: ${a[2].category}\n$Upvotes: ${a[2].upvotes}`);console.log(`\nCID: ${a[3].cid}\n$Title: ${a[3].title}\n$details: ${a[3].details}\n$category: ${a[3].category}\n$Upvotes: ${a[3].upvotes}`);console.log(`\nCID: ${a[0].cid}\n$Title: ${a[0].title}\n$details: ${a[4].details}\n$category: ${a[4].category}\n$Upvotes: ${a[4].upvotes}`)})

console.log(`\nCID: ${a[0].cid}\n$Title: ${a[0].title}\n$details: ${a[0].details}\n$category: ${a[0].category}\n$Upvotes: ${a[0].upvotes}`);console.log(`CID: ${a[1].cid}\n$Title: ${a[1].title}\n$details: ${a[1].details}\n$category: ${a[1].category}\n$Upvotes: ${a[1].upvotes}`);


src/js/app.js
App = {
  web3Provider: null,
  contracts: {},
  account: '0x0',

  init: function() {
    return App.initWeb3();
  },

  initWeb3: function() {

  if (typeof web3 !== 'undefined') {
        // If a web3 instance is already provided by Meta Mask.
        App.web3Provider = web3.currentProvider;
        web3 = new Web3(web3.currentProvider);

      } else {
        // Specify default instance if no web3 instance provided
        App.web3Provider = new Web3.providers.HttpProvider('http://localhost:7545');
        web3 = new Web3(App.web3Provider);

      }

      return App.initContract();
},

initContract: function() {

      $.getJSON("Cryptobeacon.json", function(cryptobeacon) {
        // Instantiate a new truffle contract from the artifact
        App.contracts.CBeacon = TruffleContract(cryptobeacon);
        // Connect provider to interact with contract
        App.contracts.CBeacon.setProvider(App.web3Provider);

        App.listenForEvents();
        return App.render();
      });

},

listenForEvents: function() {
  App.contracts.CBeacon.deployed().then(function(instance){
    instance.newActivityEvent({},{
      fromBlock: 0,
      toBlock: 'latest'
    }).watch(function(error,event){
      console.log("event triggered",event);
      //Reload when a new vote is recorded
      App.render();
    })
  })
},

render: function() {

  var cbeaconinstance;
  content.show();
  // Load account data
  web3.eth.getCoinbase(function(err, account) {
    if (err === null) {
      App.account = account;
      $("#accountAddress").html("Your Account: " + account);
    }
  });

  // Load contract data
  App.contracts.CBeacon.deployed().then(function(instance) {
    cbeaconinstance = instance;
    return cbeaconinstance.getLast5Activities();
  }).then(function(candidatesCount) {
    var candidatesResults = $("#candidatesResults");
    candidatesResults.empty();

    var candidatesSelect = $('#candidatesSelect');
    candidatesSelect.empty();

    for (var i = 1; i <= candidatesCount; i++) {
      electionInstance.candidates(i).then(function(candidate) {
        var id = candidate[0];
        var name = candidate[1];
        var voteCount = candidate[2];

        // Render candidate Result
        var candidateTemplate = "<tr><th>" + id + "</th><td>" + name + "</td><td>" + voteCount + "</td></tr>"
        candidatesResults.append(candidateTemplate);

        //Render candidate ballot option
        var candidateOption = "<option value='" + id + "' >" + name + "</ option>"
        candidatesSelect.append(candidateOption);
      });
    }

    return electionInstance.voters(App.account);
  }).then(function(hasVoted){
    // Do not allow a user to vote
    if(hasVoted) {
      $('form').hide();
    }


    content.show();
  }).catch(function(error) {
    console.warn(error);
  });
},

castVote: function() {
  var candidateId = $('#candidatesSelect').val();
  App.contracts.Election.deployed().then(function(instance){
    return instance.vote(candidateId, { from: App.account } );
  }).then(function(result) {
    //wait for votes to update
    $('#content').hide();
    $('#loader').hide();
  }).catch(function(err) {
    console.error(err);
  });
 }
};

$(function() {
  $(window).load(function() {
    App.init();
  });
});


src/script.js
const userInfo = [
    {
        userName: "Albert",
        passWord: "albert0"
    },
    {
        userName: "Robert",
        passWord: "robert0"
    },
    {
        userName: "Edward",
        passWord: "edward0"
    },
    {
        userName: "William",
        passWord: "william0"
    },
    {
        userName: "Johan",
        passWord: "johan0"
    }

];

// console.log(userInfo.length);

const loginSuccessful = (userName) => {
    
    document.getElementById("login-page").style.display = "none";
    document.getElementById("login-successful").style.display = "block";
    document.getElementById("userName").innerText = userName;

}

const checkValidLogin = () => {
    const userName = document.getElementById("username").value;
    const userPass = document.getElementById("password").value;
    for(let i = 0; i < userInfo.length; i++){
        if(userName == userInfo[i].userName && userPass == userInfo[i].passWord){
            loginSuccessful(userName);
            console.log("hi");
            break;
        }
    }
}

document.getElementById("sign-in").addEventListener("click", checkValidLogin);

document.getElementById("sign-in").addEventListener("click", checkValidLogin);

truffle-config.js
/**
 * Use this file to configure your truffle project. It's seeded with some
 * common settings for different networks and features like migrations,
 * compilation and testing. Uncomment the ones you need or modify
 * them to suit your project as necessary.
 *
 * More information about configuration can be found at:
 *
 * trufflesuite.com/docs/advanced/configuration
 *
 * To deploy via Infura you'll need a wallet provider (like @truffle/hdwallet-provider)
 * to sign your transactions before they're sent to a remote public node. Infura accounts
 * are available for free at: infura.io/register.
 *
 * You'll also need a mnemonic - the twelve word phrase the wallet uses to generate
 * public/private key pairs. If you're publishing your code to GitHub make sure you load this
 * phrase from a file you've .gitignored so it doesn't accidentally become public.
 *
 */

 const HDWalletProvider = require('@truffle/hdwallet-provider');
// const infuraKey = "fj4jll3k.....";
//
 const fs = require('fs');
 const mnemonic = fs.readFileSync(".secret").toString().trim();

module.exports = {
  /**
   * Networks define how you connect to your ethereum client and let you set the
   * defaults web3 uses to send transactions. If you don't specify one truffle
   * will spin up a development blockchain for you on port 9545 when you
   * run `develop` or `test`. You can ask a truffle command to use a specific
   * network from the command line, e.g
   *
   * $ truffle test --network <network-name>
   */

  networks: {
    // Useful for testing. The `development` name is special - truffle uses it by default
    // if it's defined here and no other network is specified at the command line.
    // You should run a client (like ganache-cli, geth or parity) in a separate terminal
    // tab if you use this network and you must also set the `host`, `port` and `network_id`
    // options below to some value.
    
    development: {
     host: "127.0.0.1",     // Localhost (default: none)
     port: 7545,            // Standard Ethereum port (default: none)
     network_id: "5777",       // Any network (default: none)
    },
    matic: {
      provider: () => new HDWalletProvider(mnemonic, `https://rpc-mumbai.matic.today`),
      network_id: 80001,
      confirmations: 2,
      timeoutBlocks: 200,
      skipDryRun: true,
      gasPrice: 1,  // 20 gwei (in wei) (default: 100 gwei)
    },
    //,
    // Another network with more advanced options...
    // advanced: {
    // port: 8777,             // Custom port
    // network_id: 1342,       // Custom network
    // gas: 8500000,           // Gas sent with each transaction (default: ~6700000)
    // gasPrice: 20000000000,  // 20 gwei (in wei) (default: 100 gwei)
    // from: <address>,        // Account to send txs from (default: accounts[0])
    // websocket: true        // Enable EventEmitter interface for web3 (default: false)
    // },
    // Useful for deploying to a public network.
    // NB: It's important to wrap the provider as a function.
    // ropsten: {
    // provider: () => new HDWalletProvider(mnemonic, `https://ropsten.infura.io/v3/YOUR-PROJECT-ID`),
    // network_id: 3,       // Ropsten's id
    // gas: 5500000,        // Ropsten has a lower block limit than mainnet
    // confirmations: 2,    // # of confs to wait between deployments. (default: 0)
    // timeoutBlocks: 200,  // # of blocks before a deployment times out  (minimum/default: 50)
    // skipDryRun: true     // Skip dry run before migrations? (default: false for public nets )
    // },
    // Useful for private networks
    // private: {
    // provider: () => new HDWalletProvider(mnemonic, `https://network.io`),
    // network_id: 2111,   // This network is yours, in the cloud.
    // production: true    // Treats this network as if it was a public net. (default: false)
    // }
  },

  // Set default mocha options here, use special reporters etc.
  mocha: {
    // timeout: 100000
  },

  // Configure your compilers
  compilers: {
    solc: {
       version: "0.4.22",    // Fetch exact version from solc-bin (default: truffle's version)
      // docker: true,        // Use "0.5.1" you've installed locally with docker (default: false)
      // settings: {          // See the solidity docs for advice about optimization and evmVersion
      //  optimizer: {
      //    enabled: false,
      //    runs: 200
      //  },
      //  evmVersion: "byzantium"
      // }
    }
  },

  // Truffle DB is currently disabled by default; to enable it, change enabled: false to enabled: true
  //
  // Note: if you migrated your contracts prior to enabling this field in your Truffle project and want
  // those previously migrated contracts available in the .db directory, you will need to run the following:
  // $ truffle migrate --reset --compile-all

  db: {
    enabled: false
  }
};


src/app-backup.js
const userInfo = [
  {
    userName: "Albert",
    passWord: "albert0",
  },
  {
    userName: "Robert",
    passWord: "robert0",
  },
  {
    userName: "Edward",
    passWord: "edward0",
  },
  {
    userName: "William",
    passWord: "william0",
  },
  {
    userName: "Johan",
    passWord: "johan0",
  },
];

App = {
    web3Provider: null,
    contracts: {},
    account: '0x729e969878210f18d62093fe84c7737270e274e3',
    contract_address : '0xfe87576951D8c0d0c33Fd3053Cc586a461BD8c5E',

    init: function() {
      return App.initWeb3();
    },
  
    initWeb3: function() {
  
    if (typeof web3 !== 'undefined') {
          // If a web3 instance is already provided by Meta Mask.
          App.web3Provider = web3.currentProvider;
          web3 = new Web3(web3.currentProvider);
  
        } else {
          // Specify default instance if no web3 instance provided
          App.web3Provider = new Web3.providers.HttpProvider('http://localhost:7545');
          web3 = new Web3(App.web3Provider);
  
        }
  
        return App.initContract();
  },
  
  initContract: function() {

    var MetaCoin = require("./build/contracts/Cryptobeacon.sol");

        $.getJSON("Cryptobeacon.json", function(cryptobeacon) {
          // Instantiate a new truffle contract from the artifact
          App.contracts.CBeacon = TruffleContract(cryptobeacon);
          // Connect provider to interact with contract
          App.contracts.CBeacon.setProvider(App.web3Provider);
  
          App.listenForEvents();
          return App.render();
        });
  
  },
  
  listenForEvents: function() {
    App.contracts.CBeacon.at(contract_address).then(function(instance){
      instance.newActivityEvent({},{
        fromBlock: 0,
        toBlock: 'latest'
      }).watch(function(error,event){
        console.log("event triggered",event);
        //Reload when a new vote is recorded
        App.render();
      })
    })
  },
  
  render: function() {
  
    var cbeaconinstance;

    // Load account data
    web3.eth.getCoinbase(function(err, account) {
      if (err === null) {
        App.account = account;
        $("#accountAddress").html("Your Account: " + account);
      }
    });
  
    // Load contract data
    App.contracts.CBeacon.at(contract_address).then(function(instance) {
      cbeaconinstance = instance;
      return cbeaconinstance.getLast5Activities();
    }).then(function(value) {

        let a = value;
        console.log(`\nCID: ${a[0].cid}\n$Title: ${a[0].title}\n$details: ${a[0].details}\n$category: ${a[0].category}\n$Upvotes: ${a[0].upvotes}`);
        console.log(`\nCID: ${a[1].cid}\n$Title: ${a[1].title}\n$details: ${a[1].details}\n$category: ${a[1].category}\n$Upvotes: ${a[1].upvotes}`);
        console.log(`\nCID: ${a[2].cid}\n$Title: ${a[2].title}\n$details: ${a[2].details}\n$category: ${a[2].category}\n$Upvotes: ${a[2].upvotes}`);
        console.log(`\nCID: ${a[3].cid}\n$Title: ${a[3].title}\n$details: ${a[3].details}\n$category: ${a[3].category}\n$Upvotes: ${a[3].upvotes}`);
        console.log(`\nCID: ${a[0].cid}\n$Title: ${a[0].title}\n$details: ${a[4].details}\n$category: ${a[4].category}\n$Upvotes: ${a[4].upvotes}`)
    //   var candidatesResults = $("#candidatesResults");
    //   candidatesResults.empty();
    //   var candidatesSelect = $('#candidatesSelect');
    //   candidatesSelect.empty();

    //   for (var i = 1; i <= candidatesCount; i++) {
    //     electionInstance.candidates(i).then(function(candidate) {
    //       var id = candidate[0];
    //       var name = candidate[1];
    //       var voteCount = candidate[2];
  
    //       // Render candidate Result
    //       var candidateTemplate = "<tr><th>" + id + "</th><td>" + name + "</td><td>" + voteCount + "</td></tr>"
    //       candidatesResults.append(candidateTemplate);
  
    //       //Render candidate ballot option
    //       var candidateOption = "<option value='" + id + "' >" + name + "</ option>"
    //       candidatesSelect.append(candidateOption);
    //     });
    //   }
  
      return cbeaconinstance.voters(App.account);
    })
  },
  
//   castVote: function() {
//     var candidateId = $('#candidatesSelect').val();
//     App.contracts.Election.deployed().then(function(instance){
//       return instance.vote(candidateId, { from: App.account } );
//     }).then(function(result) {
//       //wait for votes to update
//       $('#content').hide();
//       $('#loader').hide();
//     }).catch(function(err) {
//       console.error(err);
//     });
//    }
//   };
  
const loginSuccessful = (userName) => {
  document.getElementById("login-page").style.display = "none";
  document.getElementById("login-successful").style.display = "block";
  document.getElementById("userName").innerText = userName;
};

const checkValidLogin = () => {
  const userName = document.getElementById("username").value;
  const userPass = document.getElementById("password").value;
  for (let i = 0; i < userInfo.length; i++) {
    if (userName == userInfo[i].userName && userPass == userInfo[i].passWord) {
      loginSuccessful(userName);
      console.log("hi");
      break;
    }
  }
};

const postNewActivity = () => {
    const userName = document.getElementById("username").value;
    const userPass = document.getElementById("password").value;
    for (let i = 0; i < userInfo.length; i++) {
      if (userName == userInfo[i].userName && userPass == userInfo[i].passWord) {
        loginSuccessful(userName);
        console.log("hi");
        break;
      }
    }
  };

document.getElementById("sign-in").addEventListener("click", checkValidLogin);
document.getElementById("submit-activity").addEventListener("click", checkValidLogin);
// Explore
$(document).ready(function () {
  document.getElementById("heart").onclick = function () {
    document.querySelector(".fa-gratipay").style.color = "#E74C3C";
  };
});

$(function() {
    $(window).load(function() {
      App.init();
    });
  });
  